server/attr/attributeContext.go

```diff
@@ -0,0 +1,260 @@
+		}
+	}
+
+	for k := range attrs.DoubleAttributes {
+		_, present := dictionary[k]
+		if !present {
+			return fmt.Errorf("attribute index %d is not defined in the current dictionary", k)
+		}
+	}
+
+	for k := range attrs.BoolAttributes {
+		_, present := dictionary[k]
+		if !present {
+			return fmt.Errorf("attribute index %d is not defined in the current dictionary", k)
+		}
+	}
+
+	for k, v := range attrs.TimestampAttributes {
+		_, present := dictionary[k]
+		if !present {
+			return fmt.Errorf("attribute index %d is not defined in the current dictionary", k)
+		}
+
+		if _, err := ptypes.Timestamp(v); err != nil {
+			return err
+		}
+	}
+
+	for k := range attrs.BytesAttributes {
+		_, present := dictionary[k]
+		if !present {
+			return fmt.Errorf("attribute index %d is not defined in the current dictionary", k)
+		}
+	}
+
+	return nil
+}
+
+func (ac *attributeContext) update(dictionary dictionary, attrs *mixerpb.Attributes) error {
+	// check preconditions up front and bail if there are any
+	// errors without mutating the context.
+	if err := checkPreconditions(dictionary, attrs); err != nil {
+		return err
+	}
+
+	// apply all attributes
+	if attrs.StringAttributes != nil {
+		if ac.stringAttributes == nil {
+			ac.stringAttributes = make(map[string]string)
+		}
+
+		for k, v := range attrs.StringAttributes {
+			ac.stringAttributes[dictionary[k]] = v
+		}
+	}
+
+	if attrs.Int64Attributes != nil {
+		if ac.int64Attributes == nil {
+			ac.int64Attributes = make(map[string]int64)
+		}
+
+		for k, v := range attrs.Int64Attributes {
+			ac.int64Attributes[dictionary[k]] = v
+		}
+	}
+
+	if attrs.DoubleAttributes != nil {
+		if ac.float64Attributes == nil {
+			ac.float64Attributes = make(map[string]float64)
+		}
+
+		for k, v := range attrs.DoubleAttributes {
+			ac.float64Attributes[dictionary[k]] = v
+		}
+	}
+
+	if attrs.BoolAttributes != nil {
+		if ac.boolAttributes == nil {
+			ac.boolAttributes = make(map[string]bool)
+		}
+
+		for k, v := range attrs.BoolAttributes {
+			ac.boolAttributes[dictionary[k]] = v
+		}
+	}
+
+	if attrs.TimestampAttributes != nil {
+		if ac.timeAttributes == nil {
+			ac.timeAttributes = make(map[string]time.Time)
+		}
+
+		for k, v := range attrs.TimestampAttributes {
+			t, _ := ptypes.Timestamp(v)
+			ac.timeAttributes[dictionary[k]] = t
+		}
+	}
+
+	if attrs.BytesAttributes != nil {
+		if ac.bytesAttributes == nil {
+			ac.bytesAttributes = make(map[string][]uint8)
+		}
+
+		for k, v := range attrs.BytesAttributes {
+			ac.bytesAttributes[dictionary[k]] = v
+		}
+	}
+
+	// delete requested attributes
+	for _, d := range attrs.DeletedAttributes {
+		name, present := dictionary[d]
+		if present {
+			delete(ac.stringAttributes, name)
+			delete(ac.int64Attributes, name)
+			delete(ac.float64Attributes, name)
+			delete(ac.boolAttributes, name)
+			delete(ac.timeAttributes, name)
+			delete(ac.bytesAttributes, name)
+		}
+	}
+
+	return nil
+}
+
+func (ac *attributeContext) reset() {
```

Comment:
why would we use ac.reset() instead of   newAttributeContext{}   (which should ensure that maps are initaliazed)

ac.reset implies object reuse. 
ac.reset will end up trashing all the maps contained inside it.